pkg/controller/volumegroup/controller.go

```diff
@@ -0,0 +1,259 @@
+		return nil, nil
+	}
+
+	logrus.Infof("Prepare to handle LVMVolumeGroup %s changed: %v", lvmVG.Name, lvmVG)
+
+	switch lvmVG.Spec.DesireState {
+	case diskv1.VGStateEnabled:
+		logrus.Infof("Prepare to enable LVMVolumeGroup %s", lvmVG.Name)
+		return c.updateEnabledLVMVolumeGroup(lvmVG)
+	case diskv1.VGStateDisabled:
+		// should only called manually set the VGState to Disabled
+		logrus.Infof("Prepare to disable LVMVolumeGroup %s", lvmVG.Name)
+		return c.disableLVMVolumeGroup(lvmVG)
+	}
+	return nil, nil
+}
+
+func (c *Controller) OnLVMVGRemove(_ string, lvmVG *diskv1.LVMVolumeGroup) (*diskv1.LVMVolumeGroup, error) {
+	if lvmVG == nil || lvmVG.DeletionTimestamp != nil {
+		// make sure the volume group is already deleted
+		logrus.Infof("Ensure the lvm volume group is already deleted if the lvmVG CR is nil")
+		return c.removeLVMVolumeGroup(lvmVG)
+	}
+
+	return nil, nil
+}
+
+func (c *Controller) updateEnabledLVMVolumeGroup(lvmVG *diskv1.LVMVolumeGroup) (*diskv1.LVMVolumeGroup, error) {
+	logrus.Infof("Enable LVMVolumeGroup %s", lvmVG.Name)
+
+	pvsResult, err := lvm.GetPVScanResult()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pvscan result. %v", err)
+	}
+	logrus.Debugf("pvscan result: %v", pvsResult)
+	currentDevs := map[string]string{}
+	if lvmVG.Status != nil && lvmVG.Status.Devices != nil && len(lvmVG.Status.Devices) > 0 {
+		currentDevs = lvmVG.Status.Devices
+	}
+	if maps.Equal(currentDevs, lvmVG.Spec.Devices) {
+		logrus.Info("Skip this round because the devices are not changed")
+		return nil, nil
+	}
+	lvmVGCpy := lvmVG.DeepCopy()
+	if lvmVGCpy.Status == nil {
+		lvmVGCpy.Status = &diskv1.VolumeGroupStatus{}
+	}
+	// update Devices:
+	// 1. Spec.Devices >= Status.Devices -> Add/Create PV and VG
+	// 2. Spec.Devices < Status.Devices -> Remove/Reduce PV and VG
+	if lvmVG.Status != nil && len(lvmVG.Status.Devices) == 0 {
+		logrus.Warnf("No devices found in LVMVolumeGroup %s, skip", lvmVG.Name)
+		return nil, nil
+	} else if len(lvmVG.Spec.Devices) >= len(currentDevs) {
+		// add/create pv/vg
+		logrus.Infof("Spec.Devices: %v, current devices: %v", lvmVG.Spec.Devices, currentDevs)
+		err := addOrCreatePVAndVG(lvmVG, lvmVGCpy, pvsResult, currentDevs)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// reduce/remove pv/vg
+		logrus.Infof("Spec.Devices: %v, current devices: %v", lvmVG.Spec.Devices, currentDevs)
+		err := reduceOrRemovePVAndVG(lvmVG, lvmVGCpy, pvsResult, currentDevs)
+		if err != nil {
+			return nil, err
+		}
+	}
+	vgConds := diskv1.VolumeGroupCondition{
+		Type:               diskv1.VGConditionReady,
+		Status:             corev1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+		Reason:             "Volume Group is Ready",
+		Message:            fmt.Sprintf("Volume Group is Ready with devices %v", lvmVG.Spec.Devices),
+	}
+	newConds := UpdateNetworkFSConds(lvmVGCpy.Status.VGConds, vgConds)
+	lvmVGCpy.Status.VGConds = newConds
```

Comment:
might be a wrong naming? I can't figure out the network FS relation here...

Reply:
The VM import controller has a `ImportConditions` field for example. So maybe change to this in the `VolumeGroupStatus` struct.
```suggestion
	lvmVGCpy.Status.VGConditions = newConds
```

Reply:
good catch, rename to `UpdateLVMVGsConds`